### ISSUE TITLE
[TF 2.2.0rc] Regression - the `{predict,train,test}_on_batch` trace functions with fixed batch size #34907

### ISSUE THREAD

**Issue by foxik** [tensorflow/tensorflow#38742]
See #34907 for description.

Even if the referenced issue was fixed for 2.1, it is again present in TF 2.2.0rc3.

**Comment by Saduf2019** [comment]
i am able to replicate this issue, please find the [gist here](https://colab.sandbox.google.com/gist/Saduf2019/58b082e917b16769fef0319f98b8e399/38742.ipynb)

**Comment by karmel** [comment]
Thanks @foxik for the report. We are working on a long-term fix right now, which we are hoping to get into the next 2.2rc, but, if it proves difficult, may have to wait for 2.3.

**Comment by foxik** [comment]
@karmel I perfectly understand that it is late in the release cycle. If unfixed, it would make the `*_on_batch` methods unusable for dynamic size input, but affected people can always wait for 2.3.

BTW, how come that `.fit` is not affected, i.e.
```python
inputs = tf.keras.layers.Input([None])
summed = tf.math.reduce_sum(inputs, axis=1, keepdims=True)
output = tf.keras.layers.Dense(1)(summed)
model = tf.keras.Model(inputs, output)
model.compile(optimizer=tf.optimizers.Adam(), loss=tf.losses.MeanSquaredError())
for i in range(1, 300):
    model.train_on_batch(np.ones([1, i]), np.ones([1]) * i)
```
causes retraces while
```python
def dataset():
    for i in range(1, 300):
        yield np.ones([1, i]), np.ones([i]) * i

model.fit(dataset())
```
does not? Does the dataset iteration sees that the shapes are changing and retraces the training method with dynamic shapes?

**Comment by foxik** [comment]
@karmel BTW, Google Colab is using 2.2.0rc3 and I already have students hitting this issue (we have started working with RNNs).

**Comment by omalleyt12** [comment]
@foxik Thanks for the issue!

There is a potential fix in the works, but it's unclear if it will make the release cut.

I'd actually recommend moving away from the `train_on_batch` methods when possible, and instead doing custom training logic via overriding `Model.train_step`. This feature was added in this release and gives a really nice way to write custom training logic while still getting all the benefits (distribution, callbacks, data formats, etc) of `Model.fit`. It should also be noticeably faster and won't have the retracing issue

A thread explaining how this works: https://twitter.com/fchollet/status/1250622989541838848

**Comment by foxik** [comment]
@omalleyt12 Thanks! I already went through the sources of `train_step` and I like how it improved applicability -- for example we use CRF/CTC losses quite a lot, and will now be able to use .fit, which is a great simplification, even if only for the distribution strategy.

But we have sources which we want to work with pre-2.2 TF for the time being, and `train_on_batch` not working in Collab is also an issue, so it would make our life easier if the regression did not appear. But I definitely understand that it is a trade-off.

Thank you very much for all your hard work!

**Comment by perretv** [comment]
For people looking for a way to generate predictions with variable size in 2.2.0 and avoid 1) the concatenation error raised by `model.predict` 2) tf.function retracing; simply call the `model.predict_step` function in a batch loop.

**Comment by ymodak** [comment]
This is fixed with tf-nightly version '2.2.0-dev20200508'
Refer [colab gist](https://colab.research.google.com/gist/ymodak/205ae3ad0d709a84e07c90be10b46190/38742.ipynb)
Thanks!

**Comment by goldiegadde** [comment]
@foxik closing this issue as this has been resolved and the next TF 2.3.0 release will have it. Please feel free to reopen if you run into any other issues.
